File: benchmarks/tasks/mandelbrot/Mandelbrot.java
// Data parallelism: a 2000x2000 grid of independent float work, split between
// threads by taking every Nth row.

import java.util.stream.IntStream;

public class Mandelbrot {
    static final int SIDE = 2000;
    static final int LIMIT = 100;

    static long band(int start, int step) {
        long total = 0;
        for (int j = start; j < SIDE; j += step) {
            double cy = (double) j / SIDE * 3.0 - 1.5;
            for (int i = 0; i < SIDE; i++) {
                double cx = (double) i / SIDE * 3.0 - 2.0;
                double x = 0.0, y = 0.0;
                long n = 0;
                while (n < LIMIT && x * x + y * y <= 4.0) {
                    double nx = x * x - y * y + cx;
                    y = 2.0 * x * y + cy;
                    x = nx;
                    n++;
                }
                total += n;
            }
        }
        return total;
    }

    public static void main(String[] args) {
        int workers = Runtime.getRuntime().availableProcessors();
        System.out.println(
                IntStream.range(0, workers).parallel().mapToLong(w -> band(w, workers)).sum());
    }
}

File: benchmarks/tasks/mandelbrot/mandelbrot.py
"""Data parallelism: a 2000x2000 grid of independent float work, split between
processes by taking every Nth row.

Processes, not threads: CPython holds a lock on the interpreter, so threads
would take turns. `multiprocessing` is what Python reaches for, and the cost
of starting the processes is part of what it costs."""

import multiprocessing as mp

SIDE = 2000
LIMIT = 100


def band(args):
    start, step = args
    total = 0
    for j in range(start, SIDE, step):
        cy = j / SIDE * 3.0 - 1.5
        for i in range(SIDE):
            cx = i / SIDE * 3.0 - 2.0
            x = y = 0.0
            n = 0
            while n < LIMIT and x * x + y * y <= 4.0:
                x, y = x * x - y * y + cx, 2.0 * x * y + cy
                n += 1
            total += n
    return total


def main():
    workers = mp.cpu_count()
    with mp.Pool(workers) as pool:
        print(sum(pool.map(band, [(w, workers) for w in range(workers)])))


if __name__ == "__main__":
    main()
